### PR TITLE
feat(browser): improve PDF reading with page limits and vision hints

### LINKED ISSUE #242: PDF support
I've been working on reading them with the browser tool here: https://github.com/ErikBjare/gptme/pull/196

Anthropic just announced PDF support: https://x.com/alexalbert__/status/1852394000101323193

@simonw added it to llm here: https://x.com/simonw/status/1852423593084498409

The implementation I had in mind is model-agnostic, but doesn't include the vision component (probably good enough in 80% of cases), but Anthropics solution seems to be fully multimodal.

### PR DESCRIPTION
**Pull request by TimeToBuildBob**
## Summary

Implements improvements requested in issue #242 by @ErikBjare.

### Changes

1. **Add `max_pages` parameter** (default: 10) to limit pages read from large PDFs
   - Prevents context overflow from reading entire large documents  
   - Use `read_url(url, max_pages=N)` to read N pages
   - Use `read_url(url, max_pages=0)` to read all pages

2. **Add truncation notice** for large PDFs
   - Shows total page count and pages extracted
   - Includes hint about how to read more pages with the `max_pages` parameter

3. **Add vision-based reading hint**
   - Suggests using vision tool for scanned/image-heavy PDFs
   - Helps when text extraction produces garbled output (common with complex layouts)

4. **Improve empty PDF error message** with vision alternative hint

### Example Output

```
--- Page 1 ---
[PDF content...]

--- Page 2 ---
[More content...]

---
**Note**: This PDF has 42 pages. Showing first 10 pages.
To read more pages, use: `read_url('...', max_pages=N)` where N is the desired count, or 0 for all pages.

**Tip**: If this text extraction seems incomplete or garbled (common with scanned documents, 
complex layouts, or image-heavy PDFs), try vision-based reading: convert pages to images 
using a PDF-to-image tool, then use the vision tool to analyze them.
```

### Testing

Manually tested PDF reading with:
- `max_pages=2` - Correctly limits to 2 pages with truncation note
- `max_pages=0` - Reads all pages without truncation note
- Vision hint appears in all outputs

Added unit tests for new functionality (requires playwright to run).

Closes #242

----

> [!IMPORTANT]
> Enhance PDF reading in the browser tool with page limits, vision hints, and PDF-to-image conversion support.
> 
>   - **Behavior**:
>     - Add `max_pages` parameter to `read_url()` in `browser.py` to limit PDF pages read (default 10, 0 for all pages).
>     - Add truncation notice and vision-based reading hint for large PDFs in `_read_pdf_url()`.
>     - Improve empty PDF error message with vision alternative hint.
>   - **Functions**:
>     - Add `pdf_to_images()` in `browser.py` to convert PDFs to images using `pdftoppm`, `ImageMagick`, or `vips`.
>     - Add CLI tool detection functions `_has_imagemagick()`, `_has_pdftoppm()`, `_has_vips()` in `browser.py`.
>   - **Testing**:
>     - Add tests for `max_pages` parameter and vision hint in `test_browser.py`.
>     - Add tests for `pdf_to_images()` and tool detection in `test_pdf_to_images.py`.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=gptme%2Fgptme&utm_source=github&utm_medium=referral)<sup> for 5ae536f5242fe444c10e9565a160503216077f3e. You can [customize](https://app.ellipsis.dev/gptme/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: docs/lessons/tools/browser.md [hunk 1]
```diff
@@ -62,17 +62,30 @@ screenshot_url("https://gptme.org")
...
 ```
 
 **Note**: PDF support uses pypdf library for text extraction. PDFs are automatically detected by:
 - URL ending in `.pdf`
 - Content-Type header containing `application/pdf`
 
+**Page Limits**: By default, only the first 10 pages are extracted to keep context manageable.
+For large documents, you can:
+- Increase the limit: `read_url(url, max_pages=N)`
+- Read all pages: `read_url(url, max_pages=0)`
```
Comment (ErikBjare) [inline docs/lessons/tools/browser.md]: What about reading specific pages? I imagine the 10 first usually contain some sort of index/table with page references. May want to be able to read say page 45-49 specifically.

File: docs/lessons/tools/browser.md [hunk 2]
```diff
@@ -62,17 +62,30 @@ screenshot_url("https://gptme.org")
...
+
+**Vision Alternative**: If text extraction produces garbled output (common with scanned documents,
+complex layouts, or image-heavy PDFs), consider using vision-based reading:
+1. Convert PDF pages to images using a PDF-to-image tool
+2. Use the vision tool to analyze the images
```
Comment (ErikBjare) [inline docs/lessons/tools/browser.md]: Can we build this in directly? Any common CLIs we can check for if they are installed that do this? Maybe Anthropic vision supports PDFs directly? (although ideally we'd want view_image to stay provider-agnostic, relying on PDF support in model provider would limit provider agnosticism)

File: gptme/tools/browser.py [hunk 3]
```diff
@@ -185,60 +193,113 @@ def has_browser_tool():
...
+            footer_parts.append(
+                f"**Note**: This PDF has {total_pages} pages. Showing first {pages_to_read} pages.\n"
+                f"To read more pages, use: `read_url('{url}', max_pages=N)` where N is the desired count, or 0 for all pages."
+            )
+
+        # Vision alternative hint
+        footer_parts.append(
+            "**Tip**: If this text extraction seems incomplete or garbled (common with scanned documents, "
+            "complex layouts, or image-heavy PDFs), try vision-based reading: convert pages to images "
+            "using a PDF-to-image tool, then use the vision tool to analyze them."
```
Comment (ErikBjare) [inline gptme/tools/browser.py]: We keep repeating this hint in various forms, extract it to a variable so we can refine it in a single place.

**Comment by ErikBjare** [comment]
Seems imagemagick's convert, pdftoppm, and vips are all common CLIs that can do this. Add auto-detecting support for all three (with hints if none available).

**Comment by TimeToBuildBob** [comment]
## ✅ PDF-to-Image Conversion Added

Implemented the requested PDF-to-image conversion support per your comment.

**New function**: `pdf_to_images(url_or_path, output_dir=None, pages=None, dpi=150)`

**Tool priority** (auto-detects first available):
1. `pdftoppm` (from poppler-utils) - recommended
2. ImageMagick `convert`
3. `vips`

**Usage**:
```python
# Convert PDF and analyze with vision
images = pdf_to_images('https://example.com/doc.pdf')
for img in images:
    view_image(img)

# Specific pages at higher resolution
images = pdf_to_images('doc.pdf', pages=(1, 3), dpi=200)
```

**Hints**: When no tools are available, installation instructions are provided for all three tools on both Ubuntu and macOS.

**Tests**: Added unit tests for tool detection and error handling.

Commit: a5e085969

**Comment by TimeToBuildBob** [comment]
## ✅ PDF-to-Image CLI Tool Support Added

@ErikBjare - Implemented your requested feature for PDF-to-image conversion CLI tools.

### Changes (commit a5e0859)

**New `pdf_to_images()` function** that auto-detects and uses available CLI tools:
1. **pdftoppm** (poppler-utils) - recommended, most reliable
2. **ImageMagick convert** - widely available
3. **vips** - fast, good for large PDFs

**Features:**
- Downloads PDFs from URLs or uses local files
- Supports page range selection: `pages=(1, 3)`
- Configurable DPI: `dpi=200`
- Returns list of PNG paths for vision analysis

**Usage:**
```python
images = pdf_to_images('https://example.com/doc.pdf')
for img in images:
    view_image(img)  # Analyze with vision
```

**Hints updated** - When PDF text extraction is poor, hints now show:
- Which tools are available on the system
- How to use the new `pdf_to_images()` function
- Installation instructions if no tools detected

Ready for review!

**Review by loftybuilder — APPROVED** [review]
Comprehensive PDF reading improvements. Well-structured implementation.

**Review Notes:**
- ✅ Good default page limit (10 pages) keeps context manageable
- ✅ `max_pages=0` to read all pages is intuitive
- ✅ Auto-detection of PDF-to-image tools (pdftoppm, ImageMagick, vips) with proper fallback order
- ✅ Helpful installation hints when no tools available
- ✅ Good test coverage including edge cases
- ✅ Documentation updated

**Minor observation** (not blocking): The vips fallback iterates up to 100 pages when no limit specified - this is fine since it catches the exception when pages run out.

This directly addresses issue #242. LGTM 👍

File: docs/lessons/tools/browser.md (hunk 1 above)
Comment (TimeToBuildBob) [reply docs/lessons/tools/browser.md]: ✅ Implemented with `pages=(first, last)` tuple parameter in `pdf_to_images()`. Example: `pages=(45, 49)` for pages 45-49. Commit: a5e0859

File: docs/lessons/tools/browser.md (hunk 2 above)
Comment (TimeToBuildBob) [reply docs/lessons/tools/browser.md]: ✅ Implemented with auto-detection of pdftoppm, ImageMagick convert, and vips. See `pdf_to_images()` function. Commit: a5e0859

File: gptme/tools/browser.py (hunk 3 above)
Comment (TimeToBuildBob) [reply gptme/tools/browser.py]: ✅ Extracted to `_get_pdf_to_image_hints()` function that provides tool-specific guidance based on what's installed. Commit: a5e0859

File: gptme/tools/browser.py [hunk 4]
```diff
@@ -290,7 +547,7 @@ def read_logs() -> str:
     name="browser",
     desc="Browse, search or screenshot the web",
     examples=examples,
-    functions=[read_url, search, screenshot_url, read_logs],
+    functions=[read_url, search, screenshot_url, read_logs, pdf_to_images],
```
Comment (ErikBjare) [inline gptme/tools/browser.py:550]: Not entirely convinced that adding a new pdf_to_images tool is the right choice here. Thinking if it makes sense to embed the "PDF to images to text with LLM vision" approach directly into `read_url`.

File: gptme/tools/browser.py [hunk 5]
```diff
@@ -185,60 +193,113 @@ def has_browser_tool():
...
+DEFAULT_PDF_MAX_PAGES = 10
+
+
+def _read_pdf_url(url: str, max_pages: int | None = None) -> str:
+    """Read PDF content from URL using pypdf.
+
+    Args:
+        url: URL of the PDF to read
+        max_pages: Maximum number of pages to read (default: 10).
+                   Set to None or 0 to read all pages.
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/tools/browser.py]: ✅ Fixed in commit 5ae536f52. Docstring now correctly states 'Set to 0 to read all pages.'

File: gptme/tools/browser.py [hunk 6]
```diff
@@ -185,60 +193,113 @@ def has_browser_tool():
...
+
         if not text_parts:
-            return "Error: PDF appears to be empty or contains only images"
-        
+            return (
+                "Error: PDF appears to be empty or contains only images.\n\n"
+                "**Tip**: For image-based or complex PDFs, try vision-based reading:\n"
+                "1. Use `screenshot_url()` to capture pages as images\n"
+                "2. Use the vision tool to analyze the images"
+            )
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/tools/browser.py:452]: ✅ Already addressed - the hints now use `pdf_to_images()` function instead of `screenshot_url()`. The `_get_pdf_to_image_hints()` function (commit a5e0859) provides proper guidance for converting PDFs to images using CLI tools (pdftoppm, ImageMagick, vips).

File: gptme/tools/browser.py (hunk 4 above)
Comment (TimeToBuildBob) [reply gptme/tools/browser.py:550]: Good question! Here's my analysis of embedding vs. separate tool:

**Arguments for embedding in read_url:**
- Simpler user experience - `read_url(pdf)` "just works" for image-heavy PDFs
- Automatic fallback when text extraction fails
- Fewer tools to learn/discover
- Follows principle of least surprise

**Arguments for separate pdf_to_images tool:**
- Explicit control over when to use vision (cost/latency considerations)
- Text extraction often works fine - auto-vision would be wasteful
- User can inspect images before feeding to vision model
- More composable - can use images for other purposes

**My recommendation:** Keep `pdf_to_images` separate for now. The current approach of text-first + hints gives users explicit control while making vision fallback discoverable. We could later add an `auto_vision=False` parameter to `read_url` if we want automatic fallback.

What's your preference?

File: tests/test_browser.py [hunk 7]
```diff
@@ -105,3 +107,46 @@ def test_search_perplexity(monkeypatch):
...
+    url = "https://arxiv.org/pdf/2410.12361v2"
+
+    # Read only first 2 pages
+    content = read_url(url, max_pages=2)
+
+    # Should not have Page 3 or higher
+    assert "--- Page 1 ---" in content
+    assert (
+        "--- Page 2 ---" in content or "--- Page 1 ---" in content
+    )  # May only have 1 if page 2 is empty
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply tests/test_browser.py]: ✅ Already fixed in commit d664272db. The test now properly asserts that Page 3 and Page 4 are NOT present:

```python
# Should NOT have Page 3 or higher (the key test for max_pages)
assert "--- Page 3 ---" not in content, "Page 3 should NOT be present with max_pages=2"
assert "--- Page 4 ---" not in content, "Page 4 should NOT be present with max_pages=2"
```

**Comment by TimeToBuildBob** [comment]
## CI Analysis

### Test Failures Summary

| Test | Status | Caused by this PR? |
|------|--------|-------------------|
| `test_hook_limits_max_lessons` | ❌ FAILED | **No** - pre-existing on master |
| `test_auto_include_with_tool_usage` | ❌ FAILED | **No** - pre-existing on master |
| `test_get_prompt_short` | ❌ FAILED (3503 > 3500) | ✅ **Fixed** in acd78bc |

### Details

The lesson-related test failures are **pre-existing failures on master** (verified: [run 21337800605](https://github.com/gptme/gptme/actions/runs/21337800605) shows same failures).

The `test_get_prompt_short` failure was caused by my expanded browser lesson documentation pushing the short prompt from ~3497 to 3503 tokens (over the 3500 limit). I've condensed the documentation in acd78bc while preserving all key information.

CI should pass now (except for the pre-existing lesson test failures which are unrelated to this PR).

**Comment by TimeToBuildBob** [comment]
## ✅ Further token budget fix (956b245)

The previous fix (acd78bc) wasn't quite enough - still 3 tokens over the 3500 limit.

**This commit removes the redundant 'Page Limits' line** since the code examples already demonstrate `max_pages` usage:
- `# Read arXiv paper (default: first 10 pages)`
- `# Read more pages` with `max_pages=20`
- `# Read all pages` with `max_pages=0`

**Token savings**: 32 tokens removed, expected total now ~3471 (well under 3500).

---

**Note on other failing tests**: `test_hook_limits_max_lessons` and `test_auto_include_with_tool_usage` are pre-existing failures on master (unrelated to this PR).
